test/e2e/customprovider/provider.go

```diff
@@ -227,3 +233,74 @@ func (i InventoryCustomType) Store(objs object.ObjMetadataSet, status []actuatio
 func (i InventoryCustomType) GetObject() (*unstructured.Unstructured, error) {
 	return i.inv, nil
 }
+
+// Apply is an Inventory interface function implemented to apply the inventory
+// object.
+func (i InventoryCustomType) Apply(dc dynamic.Interface, mapper meta.RESTMapper, _ inventory.StatusPolicy) error {
+	invInfo, namespacedClient, err := i.getNamespacedClient(dc, mapper)
+	if err != nil {
+		return err
+	}
+
+	// Get cluster object, if exsists.
+	clusterObj, err := namespacedClient.Get(context.TODO(), invInfo.GetName(), metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	var appliedObj *unstructured.Unstructured
+
+	if clusterObj == nil {
+		// Create cluster inventory object, if it does not exist on cluster.
+		appliedObj, err = namespacedClient.Create(context.TODO(), invInfo, metav1.CreateOptions{})
+	} else {
+		// Update the cluster inventory object instead.
+		appliedObj, err = namespacedClient.Update(context.TODO(), invInfo, metav1.UpdateOptions{})
+	}
+	if err != nil {
+		return err
+	}
+
+	// Update status.
+	invInfo.SetResourceVersion(appliedObj.GetResourceVersion())
+	_, err = namespacedClient.UpdateStatus(context.TODO(), invInfo, metav1.UpdateOptions{})
+	return err
+}
+
+func (i InventoryCustomType) ApplyWithPrune(dc dynamic.Interface, mapper meta.RESTMapper, _ inventory.StatusPolicy, _ object.ObjMetadataSet) error {
+	invInfo, namespacedClient, err := i.getNamespacedClient(dc, mapper)
+	if err != nil {
+		return err
+	}
+
+	// Update the cluster inventory object.
+	appliedObj, err := namespacedClient.Update(context.TODO(), invInfo, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Update status.
+	invInfo.SetResourceVersion(appliedObj.GetResourceVersion())
+	_, err = namespacedClient.UpdateStatus(context.TODO(), invInfo, metav1.UpdateOptions{})
+	return err
+}
+
+func (i InventoryCustomType) getNamespacedClient(dc dynamic.Interface, mapper meta.RESTMapper) (*unstructured.Unstructured, dynamic.ResourceInterface, error) {
+	invInfo, err := i.GetObject()
+	if err != nil {
+		return nil, nil, err
+	}
+	if invInfo == nil {
+		return nil, nil, fmt.Errorf("attempting to create a nil inventory object")
+	}
+
+	mapping, err := mapper.RESTMapping(invInfo.GroupVersionKind().GroupKind(), invInfo.GroupVersionKind().Version)
```

Comment:
We do specify a [version](https://github.com/kubernetes-sigs/cli-utils/blob/b00a8f8d4553c9c81ee9a162905fab488a563a8f/test/e2e/customprovider/provider.go#L99). TThe RESTMapping uses a KindPriority for matching, so even if we don't specify a version (eg: `""`), it should still work.

Reply:
The way this is written implies that the version is required. I guess a better questions is: should the version be required?

It's weird to pass an empty string to a vararg. It translates to `[]string{""}` instead of `nil`, which may have subtle consequences, especially because there's half a dozen different implementations of the RESTMapper that impliment that method.

For example, the DefaultRESTMapper passed the `versions` field directly into an error when there are no mappings. The resulting error would be different if you supplied the empty string vs no arg.
https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/restmapper.go#L450

It does look like does actually skip empty string versions, but I'm not sure that all implementations of the interface do...
https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/restmapper.go#L473

So what I'm getting at is that if we require the version, we should return an error when it doesn't exist. But if it's optional, we should have an if/else so that we're not passing an empty string into the mapper.RESTMapping function and assuming all impls know how to deal with that.

But I don't know if we should require the version or not for this class/method. 

I do see that this same pattern exists in the ClusterClient.getMapping impl as well, but in that case it's used for creating/updating/deleting an object, which MUST have a version defined, otherwise the API call would fail. 

Looking at this particular usage, I see the invInfo is being build from inv.GetObject(). So I guess the question is: does GetObject always return an InvInfo with a version? But GetObject() is a method on the Storage interface, which makes that harder to answer, because not all implementations are in this code base. kpt has its own, for example.

Perhaps all we need to do is add a comment to the Storage.GetObject method and say that the returned object MUST contain Group, Version, and Kind. That way the contract is at least documented, if not enforced.

Reply:
The function `getNamespacedClient` is a helper function for `InventoryCustomType`, and is ultimately implemented by clients of cli-utils should they choose to implement the inventory as a different type on cluster. As `GetObject()` is also implemented by the client of cli-utils, it is ultimately up to the author to ensure both GetObject() returns a version and that the namespaced client created uses the right version - as is the case in this custom implementation for the e2e custom inventory type.

The namespaced client helper function is used here in cli-utils as this code branch is used within both `Apply` and `ApplyWithPrune`.